### PR TITLE
[dagster-aws] extend main ruff config + move boto3-stubs to `stubs` extras

### PR DESCRIPTION
**Pull request by danielgafni**
## Summary & Motivation

1. We currently have `boto3-stubs-lite` in `dagster-aws` runtime dependencies. This might be causing type-checkers to fail for our users if they have "bad" `boto3` code (which previously wasn't checked because no stubs package was installed). This PR moves `boto3-stubs-lite` to a new `stubs` extras of `dagster-aws`, ensuring it won't be installed by default. It's now only imported under `if TYPE_CHECKING` blocks. Packages like `mypy_boto3_glue` were added to banned top-level modules (via ruff rule). 

2. Additionally, `dagster-aws`'s ruff config now extends the root config. This mostly fixed import ordering, previously `dagster_aws` imports weren't coming after all other packages. 

## Changelog [New | Bug | Docs]

> NOCHANGELOG

### CODE REVIEW

**Comment by danielgafni** [comment]
* **#24153** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/24153?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* `master`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @danielgafni and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

**Comment by danielgafni** [comment]
`pyright` failing because of broken `master`

**Review by schrockn — CHANGES_REQUESTED** [review]
Hmm this seems complicated. 

The errors that users would see are not phantom errors, correct? They would be real issues?

I don't think this is an unreasonable change, and users can easily ignore it with `type: ignore`. 

Is this in response to user feedback?

**Comment by danielgafni** [comment]
> The errors that users would see are not phantom errors, correct? They would be real issues?

Well, kinda, but not exactly. But there are 2 arguments against them: 
- This will break existing builds for `dagster-aws` users once they update to a newer version. The errors will be popping out in random places, anywhere where `boto3` is used, even outside of Dagster code. It might be confusing for users unaware of `boto3-stubs`.  
- The number of errors might be overwhelming. `boto3-stubs` is really strict (and probably that's technically correct and aligned to AWS API specs). For example, most of response fields are set to `NotRequired`, so the user must do `response.get("field")` instead of `response["field"]`. And this might be a problem, because `pyright` doesn't have a way to automatically insert `# type: ignore` comments (if you remember, I made a script for that). 

In conclusion, this doesn't seem right to me, it should be a conscious decision to enable type-checking for `boto3` on the user side. We shouldn't be forcing best-practices onto users.

**Comment by schrockn** [comment]
Ah it will be break all all boto3 API calls. Yes then I agree this is too aggressive.

I just don't know if this is worth all the trouble and complexity at this point. How many people are going to go through the trouble of installing the appropriate extra? Let's discuss in our meeting today.

**Comment by schrockn** [comment]
Per offline discussion let's name the extra [stubs] and this is useful for our own internal development. Let's rename it and move forward.

**Comment by danielgafni** [comment]
@schrockn done
